### PR TITLE
feat(textarea): introduce isClearable

### LINKED ISSUE #2112: [BUG] - There are no isClearable at the Textarea component
### NextUI Version

2.2.9

### Describe the bug

The `isClearable` attribute is not available on the `Textarea`, but the documentation provides the `onClear` method.

### Your Example Website or App

_No response_

### Steps to Reproduce the Bug or Issue

none

### Expected behavior

I wish to be able to provide the same property configuration as the `Input` component.
Thanks.

### Screenshots or Videos

_No response_

### Operating System Version

macOS

### Browser

Chrome

### LINKED ISSUE #2348: [Feature Request] Add clear button to Textarea
### Is your feature request related to a problem? Please describe.

Input has clear button. Textarea should have too.

### Describe the solution you'd like

Add clear button to Textarea the same as Input has.

### Describe alternatives you've considered

No other way.

### Screenshots or Videos

_No response_

### PR DESCRIPTION
**Pull request by IsDyh01**
Closes #2348 
Closes #2112 

## 📝 Description

Currently, the input component has a clear button function, while the textarea component does not

## ⛳️ Current behavior (updates)

The textarea component does not have a clear button

## 🚀 New behavior

Add Clear button functionality for the textarea component。If there is no content in the textarea, the clear button will not be displayed. If there is content, the clear button will be displayed after the content. Click the clear button to clear the text in the textarea and call the callback function

## 💣 Is this a breaking change (Yes/No):
No

## Summary by CodeRabbit

- **New Features**
  - Added a clearable feature to the `Textarea` component, allowing users to easily clear the text using a clear button.
  - Enhanced the `Textarea` documentation with a new section detailing the clear button functionality.
  - Introduced a header wrapper property to the input component for improved layout styling.
  - Added a new `trash` icon to the shared icons library.

- **Bug Fixes**
  - Updated the `Textarea Props` documentation to include the new `isClearable` property.

- **Chores**
  - Updated the peer dependency for the `@nextui-org/theme` package to require a newer version.

### CODE REVIEW

**Comment by abhinandan-verma** [comment]
@IsDyh01  kindly run `pnpm changeset`  command  before commit to add the `changeset` file  then again add, commit push

**Review by wingkwong — CHANGES_REQUESTED** [review]
1. please add changeset
2. please add test

File: .changeset/five-adults-protect.md
Comment (wingkwong) [inline .changeset/five-adults-protect.md:2]: use `minor` instead

File: .changeset/five-adults-protect.md
Comment (wingkwong) [inline .changeset/five-adults-protect.md:5]: add issue number as well

```
Add a Clear button to the textarea component (#2348) 
```

File: packages/components/input/stories/textarea.stories.tsx
Comment (wingkwong) [inline packages/components/input/stories/textarea.stories.tsx:313]: why is this line commented out?

**Review by wingkwong — CHANGES_REQUESTED** [review]
please also include the corresponding info in documentation

File: .changeset/five-adults-protect.md
Comment (IsDyh01) [reply .changeset/five-adults-protect.md:2]: ok

File: .changeset/five-adults-protect.md
Comment (IsDyh01) [reply .changeset/five-adults-protect.md:5]: ok

File: packages/components/input/stories/textarea.stories.tsx
Comment (IsDyh01) [reply packages/components/input/stories/textarea.stories.tsx:313]: This is just to test for no other effects. I'll delete it！

**Review by wingkwong — CHANGES_REQUESTED** [review]
1. your test case failed
2. please also update documentation as well. you may refer to `Clear Button` section in Input.
3. currently the scrollbar will overlap with the clearbutton. please try to add a pr-4 -mr-4 to move the scrollbar without losing space.

**Comment by IsDyh01** [comment]
Ok, I will deal with these problems as soon as possible

**Comment by IsDyh01** [comment]
The scrollbar style seems to come with the browser, and you can't change its position, such as padding and margin

**Comment by wingkwong** [comment]
Actually it's possible if you change the padding of the outer container. However, we'll revisit the design first. Please check with the failed test case first.

**Comment by IsDyh01** [comment]
This commit fixes the test case failed issue, makes the textarea component support isClearable prop, and finally adds docs to the component

**Comment by wingkwong** [comment]
@IsDyh01 Just discussed with the designer, here's the new design.

- if clearable is true, the clear button will be placed at the top right. please notice the icon is different. see below for the svg code.
- the clear button will be horizontally aligned with the label and the scrollbar won't be overlap with it.

You may check the below screenshot for reference.

image

The clear button svg:

```ts
<svg fill="none" height="16" viewBox="0 0 16 16" width="16" xmlns="http://www.w3.org/2000/svg">
  <path
    d="M14.0071 3.8847L14.0087 3.88486C14.0648 3.8902 14.1093 3.93884 14.108 3.99818C14.0999 4.04945 14.0549 4.08666 14.0067 4.08666H14.0066H14.0064H14.0063H14.0062H14.0061H14.006H14.0059H14.0057H14.0056H14.0055H14.0054H14.0053H14.0052H14.005H14.0049H14.0048H14.0047H14.0046H14.0045H14.0044H14.0042H14.0041H14.004H14.0039H14.0038H14.0037H14.0036H14.0034H14.0033H14.0032H14.0031H14.003H14.0029H14.0028H14.0027H14.0025H14.0024H14.0023H14.0022H14.0021H14.002H14.0019H14.0018H14.0017H14.0015H14.0014H14.0013H14.0012H14.0011H14.001H14.0009H14.0008H14.0007H14.0005H14.0004H14.0003H14.0002H14.0001H14H13.9999H13.9998H13.9997H13.9996H13.9995H13.9993H13.9992H13.9991H13.999H13.9989H13.9988H13.9987H13.9986H13.9985H13.9984H13.9983H13.9982H13.9981H13.9979H13.9978H13.9977H13.9976H13.9975H13.9974H13.9973H13.9972H13.9971H13.997H13.9969H13.9968H13.9967H13.9966H13.9965H13.9964H13.9962H13.9961H13.996H13.9959H13.9958H13.9957H13.9956H13.9955H13.9954H13.9953H13.9952H13.9951H13.995H13.9949H13.9948H13.9947H13.9946H13.9945H13.9944H13.9943H13.9942H13.9941H13.9939H13.9938H13.9937H13.9936H13.9935H13.9934H13.9933H13.9932H13.9931H13.993H13.9929H13.9928H13.9927H13.9926H13.9925H13.9924H13.9923H13.9922H13.9921H13.992H13.9919H13.9918H13.9917H13.9916H13.9915H13.9914H13.9913H13.9912H13.9911H13.991H13.9909H13.9908H13.9907H13.9906H13.9905H13.9904H13.9903H13.9902H13.9901H13.99H13.9899H13.9898H13.9897H13.9896H13.9895H13.9894H13.9893H13.9892H13.9891H13.989H13.9889H13.9888H13.9887H13.9886H13.9885H13.9884H13.9883H13.9882H13.9881H13.988H13.9879H13.9878H13.9877H13.9876H13.9875H13.9874H13.9873H13.9872H13.9871H13.987H13.9869H13.9868H13.9867H13.9866H13.9865H13.9864H13.9863H13.9862H13.9861H13.986H13.9859H13.9858H13.9857H13.9856H13.9855H13.9854H13.9853H13.9852H13.9851H13.985H13.9849H13.9848H13.9847H13.9846H13.9845H13.9844H13.9843H13.9842H13.9841H13.984H13.9839H13.9838H13.9837H13.9836H13.9835H13.9834H13.9833H13.9832H13.9831H13.983H13.9829H13.9828H13.9827H13.9826H13.9825H13.9824H13.9823H13.9823H13.9822H13.9821H13.982H13.9819H13.9818H13.9817H13.9816H13.9815H13.9814H13.9813H13.9812H13.9811H13.981H13.9809H13.9808H13.9807H13.9806H13.9805H13.9804H13.9803H13.9802H13.9801H13.98H13.9799H13.9798H13.9797H13.9796H13.9795H13.9794H13.9793H13.9792H13.9791H13.979H13.9789H13.9788H13.9787H13.9786H13.9785H13.9784H13.9783H13.9782H13.9781H13.9781H13.978H13.9779H13.9778H13.9777H13.9776H13.9775H13.9774H13.9773H13.9772H13.9771H13.977H13.9769H13.9768H13.9767H13.9766H13.9765H13.9764H13.9763H13.9762H13.9761H13.976H13.9759H13.9758H13.9757H13.9756H13.9755H13.9754H13.9753H13.9752H13.9751H13.975H13.9749H13.9748H13.9747H13.9746H13.9745H13.9744H13.9743H13.9742H13.9741H13.974H13.9739H13.9738H13.9737H13.9736H13.9735H13.9734H13.9733H13.9733C10.4314 3.7333 6.88973 3.59909 3.36755 3.94858C3.36742 3.94859 3.36729 3.9486 3.36717 3.94862L2.00875 4.0818C2.00861 4.08181 2.00848 4.08182 2.00835 4.08183C1.94717 4.08747 1.90326 4.04725 1.89811 3.99449C1.8929 3.9411 1.93182 3.89052 1.99126 3.88486L1.99237 3.88475L3.35187 3.75147C3.35194 3.75146 3.35201 3.75145 3.35208 3.75145C3.96457 3.69218 4.57811 3.65256 5.19909 3.6125L5.51656 3.59202L5.56806 3.27809L5.708 2.42509C5.76399 2.08568 5.81767 1.81209 5.98475 1.60624C6.12965 1.42771 6.42122 1.23333 7.12 1.23333H8.86667C9.56355 1.23333 9.85752 1.43445 10.0046 1.61915C10.1739 1.83176 10.2261 2.10927 10.2786 2.43103L10.2785 2.43103L10.2789 2.4334L10.42 3.26694V3.61839L10.7978 3.63938C11.8728 3.6991 12.9403 3.77868 14.0071 3.8847Z"
    fill="#A1A1AA"
    stroke="#A1A1AA"
    strokeWidth="0.8"
  />
  <path
    d="M12.2134 5.56667C12.3292 5.56667 12.4438 5.61405 12.5285 5.70064C12.6088 5.78957 12.6542 5.90963 12.6474 6.03693C12.6474 6.03722 12.6473 6.03752 12.6473 6.03782L12.2342 12.8669C12.1959 13.3993 12.1525 13.8463 11.9087 14.1798C11.69 14.4792 11.2352 14.7733 10.14 14.7733H5.86004C4.76563 14.7733 4.31057 14.4779 4.09155 14.1776C3.8475 13.843 3.80414 13.3957 3.76583 12.8668L3.35272 6.03052C3.35271 6.03034 3.3527 6.03016 3.35269 6.02998C3.34597 5.90978 3.39108 5.78807 3.47441 5.69749C3.54714 5.61844 3.66453 5.56667 3.7867 5.56667H12.2134ZM6.8867 12.2333H9.1067C9.60095 12.2333 10.0067 11.8276 10.0067 11.3333C10.0067 10.8391 9.60095 10.4333 9.1067 10.4333H6.8867C6.39246 10.4333 5.9867 10.8391 5.9867 11.3333C5.9867 11.8276 6.39246 12.2333 6.8867 12.2333ZM6.33337 9.56667H9.6667C10.161 9.56667 10.5667 9.16092 10.5667 8.66667C10.5667 8.17242 10.161 7.76667 9.6667 7.76667H6.33337C5.83912 7.76667 5.43337 8.17242 5.43337 8.66667C5.43337 9.16092 5.83912 9.56667 6.33337 9.56667Z"
    fill="#A1A1AA"
    stroke="#A1A1AA"
    strokeWidth="0.8"
  />
</svg>;
```

**Comment by IsDyh01** [comment]
ok，I will modify it as soon as possible.

**Comment by IsDyh01** [comment]
currently behavior 
image
image

**Comment by wingkwong** [comment]
@IsDyh01 the case without label fails

image

File: packages/components/input/src/textarea.tsx
Comment (wingkwong) [inline packages/components/input/src/textarea.tsx:127]: why we need `top-2` here?

File: packages/utilities/shared-icons/src/dustbin.tsx
Comment (wingkwong) [inline packages/utilities/shared-icons/src/dustbin.tsx:3]: rename to `TrashIcon` since it's from `vuesax`. Also rename the file name.

**Review by wingkwong — CHANGES_REQUESTED** [review]

File: packages/components/input/src/textarea.tsx
Comment (IsDyh01) [reply packages/components/input/src/textarea.tsx:127]: because the clear button is absolutely positioned on the outermost div element and its original behavior is: 
image
so i have to add **top-2** classname to position it at the top:
image

File: packages/utilities/shared-icons/src/dustbin.tsx
Comment (IsDyh01) [reply packages/utilities/shared-icons/src/dustbin.tsx:3]: ok,i will change it with the case without label

File: packages/components/input/src/textarea.tsx
Comment (wingkwong) [reply packages/components/input/src/textarea.tsx:127]: i think you may revise the structure based on the new design. It shouldn't require `top-2` at all in the new design.

Comment (IsDyh01) [reply packages/components/input/src/textarea.tsx:127]: ok

Comment (IsDyh01) [reply packages/components/input/src/textarea.tsx:127]: I don't understand why this clear button has a default top position

Comment (IsDyh01) [reply packages/components/input/src/textarea.tsx:127]: sorry, could give me some tips, how to revise the structure?

Comment (IsDyh01) [reply packages/components/input/src/textarea.tsx:127]: sorry, could u give me some tips, how to revise the structure?

Comment (wingkwong) [reply packages/components/input/src/textarea.tsx]: label and clear button will be on the same row. currently clear button is using absolute which is no longer required.

Comment (IsDyh01) [reply packages/components/input/src/textarea.tsx]: thanks，i know how to do it.

Comment (IsDyh01) [reply packages/components/input/src/textarea.tsx]: image
image

File: packages/components/input/src/textarea.tsx
Comment (wingkwong) [inline packages/components/input/src/textarea.tsx:156]: move the style classes to `getInputWrapperProps`

File: packages/components/input/src/textarea.tsx
Comment (wingkwong) [inline packages/components/input/src/textarea.tsx:128]: move the style classes to `getClearButtonProps `. also negative margin shouldn't be used.

**Review by wingkwong — CHANGES_REQUESTED** [review]

File: packages/components/input/src/textarea.tsx
Comment (IsDyh01) [reply packages/components/input/src/textarea.tsx:156]: if move the style to getInputWrapperProps ,the input component is also affected

File: packages/components/input/src/textarea.tsx
Comment (IsDyh01) [reply packages/components/input/src/textarea.tsx:128]: the negative margin is to align the clear button with the label and scroll bar

File: packages/components/input/src/textarea.tsx
Comment (wingkwong) [reply packages/components/input/src/textarea.tsx:156]: there is a variable called `isMultiline`. if it's `true`, then it's textarea, else input.

File: packages/components/input/src/textarea.tsx
Comment (wingkwong) [reply packages/components/input/src/textarea.tsx:128]: please remove those negative margins. that is not the correct way to align the clear button with the scrollbar.

**Review by wingkwong — CHANGES_REQUESTED** [review]
1. please double check the space above label. it should align with the existing production one.

your pr
image

production
image

2. unexpected padding

image

3. please check [RTL case](https://nextui-storybook-v2-git-fork-isdyh01-feat-tex-2b545f-nextui-org.vercel.app/?path=%2Fstory%2Fcomponents-textarea--clearable&globals=locale%3Aar-AE)
image

**Comment by IsDyh01** [comment]
image
image
image

File: packages/components/input/src/use-input.ts
Comment (wingkwong) [inline packages/components/input/src/use-input.ts:527]: why we need these classes? feel like none is required.

File: packages/components/input/src/use-input.ts
Comment (wingkwong) [inline packages/components/input/src/use-input.ts:373]: seems not required

**Review by wingkwong — CHANGES_REQUESTED** [review]
image

File: packages/components/input/src/use-input.ts
Comment (IsDyh01) [reply packages/components/input/src/use-input.ts:373]: pr-0 is because pr-6 is added to the input or textarea element if isClearable is true, which is the unexpected padding problem you mentioned above

File: packages/components/input/src/use-input.ts
Comment (IsDyh01) [reply packages/components/input/src/use-input.ts:527]: Make the button relatively positioned and have its position placed in the correct place

Comment (wingkwong) [reply packages/components/input/src/use-input.ts:531]: if i delete `relative block opacity-100 p-0 -m-0 end-0`, the clear button is still in the same place.

File: packages/components/input/src/use-input.ts
Comment (wingkwong) [inline packages/components/input/src/use-input.ts:376]: we don't use `rtl:`. please sync once with canary branch then use `pe-0`.

**Review by wingkwong — CHANGES_REQUESTED** [review]

File: packages/components/input/src/use-input.ts
Comment (IsDyh01) [reply packages/components/input/src/use-input.ts:376]: pe-0 doesn't seem to be working，and **rtl:** mainly solves the padding problem in PTL case

Comment (wingkwong) [reply packages/components/input/src/use-input.ts:376]: you need to sync canary branch to your working branch first because your branch is still using `rtl:` so `pe-0` won't work.

File: packages/components/input/src/use-input.ts
Comment (IsDyh01) [reply packages/components/input/src/use-input.ts:531]: That's because I've already used these styles to override the old ones, so the position won't change if I remove these attributes

Comment (IsDyh01) [reply packages/components/input/src/use-input.ts:531]: image
These are the default styles

File: packages/components/input/src/use-input.ts
Comment (IsDyh01) [reply packages/components/input/src/use-input.ts]: thanks，I've revise it

File: packages/components/input/src/use-input.ts
Comment (wingkwong) [inline packages/components/input/src/use-input.ts:376]: no need `px-0` ..

**Review by wingkwong — CHANGES_REQUESTED** [review]

File: packages/components/input/src/use-input.ts
Comment (IsDyh01) [reply packages/components/input/src/use-input.ts]: ok，i have revise it

File: packages/components/input/src/textarea.tsx
Comment (wingkwong) [inline packages/components/input/src/textarea.tsx:150]: don't define classes here. follow the pattern - use `getXXXXProps()`.

File: packages/components/input/__tests__/textarea.test.tsx
Comment (wingkwong) [inline packages/components/input/__tests__/textarea.test.tsx:7]: add one more test case. define `onClear` without defining `isClearable`. It should have clear button as well.

**Review by wingkwong — CHANGES_REQUESTED** [review]
final round of comments. please pull the code once first.

File: packages/components/input/__tests__/textarea.test.tsx
Comment (IsDyh01) [reply packages/components/input/__tests__/textarea.test.tsx:7]: OK

File: packages/components/input/src/textarea.tsx
Comment (IsDyh01) [reply packages/components/input/src/textarea.tsx:150]: OK

**Review by wingkwong — CHANGES_REQUESTED** [review]
it breaks the existing layout - With End Content. Please double check all the cases.

prod:
image

your pr:
image

**Comment by IsDyh01** [comment]
can endContent and clearButton appear at the same time in the textarea？endContent has a higher priority in the input.

File: packages/components/input/src/use-input.ts
Comment (wingkwong) [inline packages/components/input/src/use-input.ts:539]: use slot. please check other getXXXXProps for reference. update docs as well.

**Review by wingkwong — CHANGES_REQUESTED** [review]
check coderabbitai comments as well

File: .changeset/five-adults-protect.md
Comment (wingkwong) [inline .changeset/five-adults-protect.md:3]: also include theme package since you've changed it.

**Review by wingkwong — CHANGES_REQUESTED** [review]

File: packages/components/input/src/textarea.tsx
Comment (IsDyh01) [reply packages/components/input/src/textarea.tsx:126]: the isClearable is not the pass props, its obtained from useInput, and if just set onClear, the isClearable from useInput is also true

File: packages/core/theme/src/components/input.ts
Comment (IsDyh01) [reply packages/core/theme/src/components/input.ts:72]: sorry, could tell me how to update this files or can you change it for me？

Comment (IsDyh01) [reply packages/core/theme/src/components/input.ts:72]: This change does not seem to affect the docs

Comment (wingkwong) [reply packages/core/theme/src/components/input.ts:72]: for the first point, because you added a new slot `headerWrapper`, so you need to document it. See here: https://nextui.org/docs/components/textarea#slots

for the second point, since you added something new to input theme, so you need to update the min peer dependency.

**Review by wingkwong — CHANGES_REQUESTED** [review]
https://github.com/nextui-org/nextui/pull/3477#discussion_r1807663263

**Comment by IsDyh01** [comment]
I'm very sorry for the late revision. I had something to do in the last two weeks

File: packages/components/input/package.json
Comment (wingkwong) [inline packages/components/input/package.json:39]: current theme version is 2.2.11. if you apply minor on theme, then it should be `>=2.3.0` instead.

**Review by wingkwong — CHANGES_REQUESTED** [review]

**Comment by IsDyh01** [comment]
I've merged the latest code

**Comment by IsDyh01** [comment]
@wingkwong I've merged the latest code.

File: .changeset/five-adults-protect.md
Comment (jrgarciadev) [inline .changeset/five-adults-protect.md:5]:
```suggestion
"@nextui-org/input": patch
"@nextui-org/shared-icons": patch
"@nextui-org/theme": patch
---
```

**Comment by IsDyh01** [comment]
> @IsDyh01 the clear button should be visible only when the textarea has some content
> 
> CleanShot 2024-11-05 at 08 18 28
> 
> please take a look at the input behavior https://nextui.org/docs/components/input#clear-button

I already took care of it

**Comment by wingkwong** [comment]
mark it on hold until the doc structure revamp pr got merged to beta branch

File: apps/docs/content/docs/components/textarea.mdx
Comment (jrgarciadev) [inline apps/docs/content/docs/components/textarea.mdx:101]:
```suggestion
<CodeDemo title="Controlled" files={textareaContent.controlled} />

> **Note**: NextUI `Textarea` also supports native events like `onChange`, useful for form libraries
> such as [Formik](https://formik.org/) and [React Hook Form](https://react-hook-form.com/).

### Clear Button

If you pass the `isClearable` property to the textarea, it will have a clear button at the
end of the textarea, it will be visible when the textarea has a value.

<CodeDemo title="Clear Button" files={textareaContent.clearButton} />
```

File: packages/components/input/src/use-input.ts
Comment (jrgarciadev) [inline packages/components/input/src/use-input.ts:418]: @IsDyh01 Tailwind classes should be exclusively defined in the theme package, specifically in the `input.ts` file. This is because the TailwindCSS content path is configured to target only the `theme` package

e.g.
```
// tailwind.config.js
import {nextui} from "@nextui-org/react";

/** @type {import('tailwindcss').Config} */
const config = {
  content: [
    // ...
    // make sure it's pointing to the ROOT node_module
    "./node_modules/@nextui-org/theme/dist/**/*.{js,ts,jsx,tsx}" // <- theme package
  ],
  theme: {
    extend: {},
  },
  darkMode: "class",
  plugins: [nextui()]
}

export default config;
```

File: packages/utilities/shared-icons/src/index.ts
Comment (jrgarciadev) [inline packages/utilities/shared-icons/src/index.ts:40]: @IsDyh01 Delete this icon since we're not using it

**Comment by jrgarciadev** [comment]
@IsDyh01 I added some comments

**Comment by IsDyh01** [comment]
ok, I will modify it as soon as possible.

**Comment by jrgarciadev** [comment]
Hey @IsDyh01 thanks for this contribution, however, I'm closing this PR in favor of this one https://github.com/nextui-org/nextui/pull/4172 (I had to some a couple of fixes)

**Pull request closed without merging**
